tools/lib/url_file.py

```diff
@@ -34,9 +47,65 @@ def __exit__(self, exc_type, exc_value, traceback):
       self._local_file = None
 
   @retry(wait=wait_random_exponential(multiplier=1, max=5), stop=stop_after_attempt(3), reraise=True)
+  def get_length(self):
+    if self._length is not None:
+      return self._length
+    path_to_name = PATH + LENGTHS_SUBFOLDER + str(sha256((self._url.split("?")[0]).encode('utf-8')).hexdigest()) + "_length"
```

Comment:
- Don't put lengths in a subfolder
 - Use `os.path.join` to add subpath folders together.

Reply:
Abstract url hashing into method. It's repeated below